src/target/ext/host_io.rs

```diff
@@ -0,0 +1,19 @@
+//! Provide Host I/O operations for the target.
+use crate::target::Target;
+
+/// Target Extension - Perform I/O operations on host
+pub trait HostIo: Target {
+    /// Open a file at filename and return a file descriptor for it, or return
+    /// -1 if an error occurs.
+    fn open(&self, filename: &[u8], flags: u64, mode: u64) -> i64;
+    /// Close the open file corresponding to fd and return 0, or -1 if an error
+    /// occurs.
+    fn close(&self, fd: usize) -> i64;
+    /// Read data from the open file corresponding to fd.
+    fn pread(&self, fd: usize, count: usize, offset: usize) -> &[u8];
```

Comment:
I don't think we should be returning a `&[u8]` here. The lifetime of that buffer will be tied to `&self`, which would make implementing this method for non-static strings a bit annoying. e.g: it would require `self` to have a `host_io_buf: Vec<u8>` field into which data is written, and a handle is passed back.

Worse still, on no_std platforms where `Vec<u8>` isn't available, it might not even be _possible_ to implement this method, as there wouldn't be any way to "steam" data directly from a file back to GDB.

Instead, I would swap out the API for something like:

```suggestion
    fn pread(&self, fd: usize, count: usize, offset: usize, output: &mut PreadOutput<'_>);
```

where `PreadOutput` is a wrapper around something like:

```rust
struct PreadOutput<'a> {
    cb: &'a mut dyn FnMut(&[u8]);
}

impl<'a> PreadOutput<'a> {
    fn write(&mut self, buf: &[u8]) { (self.cb)(buf) }
}
```

* * *

Essentially, you'd want to switch to a callback-based interface, whereby the implementation doesn't have to allocate a buffer tied to the lifetime of `self` just to send data back.

Take a look at the implementation of `MonitorCmd` and `ConsoleOutput` for a better idea of how this would work.

Reply:
Where should the definition of `PreadOutput` be placed at?

And now we have `ConsoleOutput`, `PreadOutput` and `SendRegisterOutput` in #58. Could we union them into a general structure?

Reply:
You can define `PreadOutput` under `target::ext::host_io::`

Yeah... those structures all share the same internals, don't they? They should definitely stay as separate types for logical consistency, but it might make sense to unify the underlying impls.

In any case, that's something best left to a follow up PR / cleanup commit.

Reply:
Hope for details